feat/nativead

alhiwatan opened the pull request:
### Initial review for supporting Native ads

NativeAdView is a container view for supporting the AppLovin Max native ads.   NativeAdView can contain the native ad components together with your components.   NativeAdView loads a native ad over the Internet and renders it with the native ad components.

```
    <AppLovinMAX.NativeAdView adUnitId={NATIVE_AD_UNIT_ID}
                              placement='myplacement'
                              customData='mycustomdata'
                              ref={nativeAdViewRef}
                              style={styles.nativeadview}>
```

The native ad components include IconView, TitleVIew, AdvertiserView, BodyView, MediaView, OptionsView, and CallToActionView.    The standard React Native properties apply to the native ad components for you to layout, change colors, and so on.

```
    <AppLovinMAX.NativeAdView adUnitId={NATIVE_AD_UNIT_ID} style={style.nativeadview} ...>
        <AppLovinMAX.NativeAdView.IconView style={styles.icon}/>
        <AppLovinMAX.NativeAdView.TitleView style={styles.title}/>
        <AppLovinMAX.NativeAdView.AdvertiserView style={styles.advertiser}/>
        <AppLovinMAX.NativeAdView.BodyView style={styles.body}/>
        <AppLovinMAX.NativeAdView.MediaView style={styles.mediaView}/>
        <AppLovinMAX.NativeAdView.CallToActionView style={styles.callToAction}/>
    </AppLovinMAX.NativeAdView>
```

NativeAdView loads a native ad as soon as it is mounted. Once it is mounted, you can reload a next native ad using a ref that is provided to NativeAdView.

```
    nativeAdViewRef.current?.loadAd();
```

**Event listeners**

```
    AppLovinMAX.addEventListener('OnNativeAdLoadedEvent', (adInfo) => {
      console.log('Native ad loaded from: ' + adInfo.networkName);
    });
    AppLovinMAX.addEventListener('OnNativeAdLoadFailedEvent', (errorInfo) => {
      console.log('Native ad failed to load with error code ' + errorInfo.code + ' and message: ' + errorInfo.message);
    });
    AppLovinMAX.addEventListener('OnNativeAdClickedEvent', (adInfo) => {
      console.log('Native ad clicked');
    });
    AppLovinMAX.addEventListener('OnNativeAdRevenuePaid', (adInfo) => {
      console.log('Native ad revenue paid: ' + adInfo.revenue);
    });
```

**Ad Units**

NativeAdView supports small or medium templates.   Manual is not supported.

**Internal Architecture:**

NativeAdView is a native UI component that implements UIView for iOS and ReactViewGroup for Android in the native module, while the native ad components are the standard React Native components. 

| React Native | Native Ad Components |
----|---- 
| Text |  TitleVIew, AdvertiserView, BodyView, CallToActionView|
| View | MediaView, OptionsView |
| Image | IconView |

NativeAdView loads a native ad using the platform-specific AppLovin Max APIs in the native module.  Then, the native module converts the loaded native ad to a JSON object and sends it back to NativeAdView.   The JSON object will be propagated to the native ad components for rendering via the React context provider framework.   This is a table of the `nativeAd` JSON object.

| Field | Type| Description |
----|----|----
| title |  string | title text |
| advertiser | string | advertiser text |
| body | string | body text |
| icon | string | image URL text |
| image | boolean | true for native image view |
| callToAction | string | CTA button text |
|mediaContentAspectRatio| float | media view aspect ratio |

MediaView, OptionsView, and IconView for images are treated differently.  The AppLovin Max APIs provide a platform-specific native view for them.  NativeAdView places it on top of the corresponding React Native component.

**Example Usage**

```
const NativeAd = () => {
  const nativeAdViewRef = useRef();

  // loading a new native ad
  const onLoadPressed = () => {
    nativeAdViewRef.current?.loadAd();
  };

  return (
    <AppLovinMAX.NativeAdView adUnitId={NATIVE_AD_UNIT_ID}
                              placement='myplacement'
                              customData='mycustomdata'
                              ref={nativeAdViewRef}
                              style={styles.nativead}>
      <View style={{flex: 1, flexDirection: 'column'}}>
        <View style={{flexDirection: 'row'}}>
          <AppLovinMAX.NativeAdView.IconView style={styles.icon}/>
          <View style={{flexDirection: 'column', flexGrow: 1}}>
            <AppLovinMAX.NativeAdView.TitleView style={styles.title}/>
            <View style={{flexDirection: 'row',
                          justifyContent: 'space-between',
                          paddingHorizontal: 20}}>
              <AppLovinMAX.NativeAdView.AdvertiserView style={styles.advertiser}/>
              <Text style={styles.admark}>AD</Text>
            </View>
          </View>
        </View>
        <AppLovinMAX.NativeAdView.BodyView style={styles.body}/>
        <AppLovinMAX.NativeAdView.MediaView style={styles.mediaView}/>
        <AppLovinMAX.NativeAdView.CallToActionView style={styles.callToAction}/>
      </View>
    </AppLovinMAX.NativeAdView>
  );
};
```

thomasmso commented on example/src/App.tsx:
```diff
@@ -364,6 +382,64 @@ const App = () => {
+        </View>
+        {
+          isNativeAdShowing &&
+            <AppLovinMAX.NativeAdView adUnitId={NATIVE_AD_UNIT_ID}
+                                      placement='myplacement'
+                                      customData='mycustomdata'
+                                      extraParameter={{
+                                        'key1': 'value1',
+                                        'key2': 'value2',
+                                      }}
+                                      ref={nativeAdViewRef}
+                                      style={styles.nativead}>
+              <View style={{flex: 1, flexDirection: 'column'}}>
+                <View style={{flexDirection: 'row'}}>
+                  <AppLovinMAX.IconView style={styles.icon}/>
+                  <View style={{flexDirection: 'column', flexGrow: 1}}>
+                    <AppLovinMAX.TitleView style={styles.title}/>
+                    <View style={{flexDirection: 'row',
+                                  justifyContent: 'space-between',
+                                  paddingHorizontal: 10}}>
+                      <AppLovinMAX.AdvertiserView style={styles.advertiser}/>
+                      <Text style={styles.admark}>AD</Text>
+                    </View>
+                  </View>
+                </View>
+                <AppLovinMAX.BodyView style={styles.body}/>
+                <AppLovinMAX.MediaView style={styles.mediaView}/>
+                <AppLovinMAX.CallToActionView style={styles.callToAction}/>
+              </View>
+            </AppLovinMAX.NativeAdView>
```
Can we just break this whole component and its styling out into its own `NativeAdView`?

alhiwatan replied:
Externalized to NativeAdView.

thomasmso commented on src/NativeAdComponents.js:
```diff
@@ -0,0 +1,114 @@
+  const {nativeAd, nativeAdView} = useContext(NativeAdViewContext);
+
+  useEffect(() => {
+    if (Object.keys(nativeAd).length === 0) return;
+    if (nativeAd.image) {
```
Is `if (Object.keys(nativeAd).length === 0) return;` even necessary when you already have the `if (nativeAd.image) {` to check if there's an image?

alhiwatan replied:
Right, removed.

thomasmso commented on src/NativeAdComponents.js:
```diff
@@ -0,0 +1,114 @@
+};
+
+export const IconView = (props) => {
+  const ref = useRef();
```
We should be more descriptive with naming, this references the mounted `Image` - so can we rename to `imageRef`?

alhiwatan replied:
done

thomasmso commented on src/NativeAdComponents.js:
```diff
@@ -0,0 +1,114 @@
+};
+
+export const MediaView = (props) => {
+  const ref = useRef();
```
Same comment as above - `viewRef`.

alhiwatan replied:
done

thomasmso commented on src/NativeAdComponents.js:
```diff
@@ -0,0 +1,114 @@
+  const {nativeAd, nativeAdView} = useContext(NativeAdViewContext);
+
+  const onLayout = () => {
+    if (Platform.OS === "android") {
```
What happens on iOS?

alhiwatan replied:
Seems working fine now, so removed.

thomasmso commented on src/NativeAdComponents.js:
```diff
@@ -0,0 +1,114 @@
+  );
+};
+
+export const TitleView = (props) => {
```
nit but can we order the exports in the logical ordering of codebase:

```
public class MaxNativeAd
{
    private final MaxAdFormat      format;
    private final String           title;
    private final String           advertiser;
    private final String           body;
    private final String           callToAction;
    private final MaxNativeAdImage icon;
    private final MaxNativeAdImage mainImage;
    private final View             iconView;
    private final View             optionsView;
    private final View             mediaView;
    private final float            mediaContentAspectRatio;
    ...
```

thomasmso reviewed (commented):
In the docs - we'll ha

alhiwatan commented on src/index.js:
```diff
@@ -81,6 +83,30 @@ const showRewardedAd = (adUnitId, ...args) => {
   }
 };
 
+const preLoadNativeAd = (adUnitId, ...args) => {
```
As suggested below, preload has been removed.

alhiwatan commented on src/index.js:
```diff
@@ -1,6 +1,8 @@
 import AdView, { AdFormat, AdViewPosition } from "./AppLovinMAXAdView";
 import { TargetingData as targetingData, AdContentRating, UserGender } from "./TargetingData";
+import { NativeAdViewWrapper as NativeAdView } from "./NativeAdView";
+import { AdvertiserView, BodyView, CallToActionView, IconView, MediaView, OptionsView, TitleView } from "./NativeAdComponents";
```
I was not able to find the general convention but it seems that the cascade namespaces are preferred.   So,  will change like <AppLovinMAX.NativeAdView.IconView ...

thomasmso replied:
It is verbose, but it leaves nothing to the imagination, thanks!

alhiwatan commented on android/src/main/java/com/applovin/reactnative/AppLovinMAXModule.java:
```diff
@@ -906,6 +906,13 @@ public void setRewardedAdExtraParameter(final String adUnitId, final String key,
         rewardedAd.setExtraParameter( key, value );
     }
 
+    @ReactMethod()
+    public void preLoadNativeAd(final String adUnitId, @Nullable final String placement, @Nullable final String customData, @Nullable final ReadableMap extraParameter)
+    {
+        Map extraParameterMap = ( extraParameter != null ) ? extraParameter.toHashMap() : null;
+        AppLovinMAXNativeAdLoader.getInstance().preLoadAd( getReactApplicationContext(), adUnitId, placement, customData, extraParameterMap );
+    }
+
```
done

alhiwatan commented on android/src/main/java/com/applovin/reactnative/AppLovinMAXNativeAdLoader.java:
```diff
@@ -0,0 +1,299 @@
+        }
+    }
+
+    public static AppLovinMAXNativeAdLoader getInstance()
```
Ok, removed AppLovinMaxNativeAdLoader and moved the functions inside AppLovinMaxNativeView.

alhiwatan commented on android/src/main/java/com/applovin/reactnative/AppLovinMAXNativeAdLoader.java:
```diff
@@ -0,0 +1,299 @@
+
+    private AppLovinMAXNativeAdLoader() { }
+
+    MaxNativeAdLoader getNativeAdLoader(Context context, String adUnitId)
```
Done

alhiwatan commented on android/src/main/java/com/applovin/reactnative/AppLovinMAXNativeAdLoader.java:
```diff
@@ -0,0 +1,299 @@
+        }
+    }
+
+    void sendCallToAction(MaxAd ad)
```
This actually sends a click event to the button, so it may or may not invoke the actual click action depending on the Max implementation.

alhiwatan commented on android/src/main/java/com/applovin/reactnative/AppLovinMAXNativeAdLoader.java:
```diff
@@ -0,0 +1,299 @@
+    void preLoadAd(Context context, String adUnitId, String placement, String customData, Map extraParameter)
+    {
+        List<MaxAd> list = nativeAds.get( adUnitId );
+        if ( list != null && list.size() > MAX_PRELOADED_ADS )
```
Understood, removed prealoadAd().

alhiwatan commented on example/src/App.tsx:
```diff
@@ -220,6 +224,20 @@ const App = () => {
+
+    // Native Ad Listeners
+    AppLovinMAX.addEventListener('OnNativeAdLoadedEvent', (adInfo) => {
+      setStatusText('NATIVE ad loaded from: ' + adInfo.networkName);
```
Renamed to Native

alhiwatan commented:
@thomasmso thanks a lot for the review.   Since I've updated a lot, I forcibly overwrote the branch for clean review.    The update includes:

- Removed AppLovinMAXNativeAdLoader - no singleton, no preload, no expiration logic
- Re-namespaced the child Ad components under AppLovin.NativeAdView, such as <AppLovinMAX.NativeAdView.IconView/>
- Added NativeAdView to the demo app

Regarding the click button event, it simulates the user click event programmatically via the nativead's CTA button, so it does not call MANativeAd.performClick().

thomasmso commented:
I need to allocate some time to properly review this, but one architectural item is the ad loader logic should be decoupled from the native ad view, and the NativeAdView should contain view-related code.

thomasmso commented on ios/AppLovinMAXNativeAdView.m:
```diff
@@ -0,0 +1,258 @@
+    }
+}
+
+- (void)load:(NSString *)adUnitIdentifier :(NSString *)placement :(NSString *)customData :(NSDictionary *)extraParameters
```
How come no need for `RCT_EXPORT_METHOD` ? Would be nice to have some sort of indicator that this can be called form react layer.

alhiwatan replied:
RCT_EXPORT_METHOD is defined on the manager side.   On the view side, the function is called from the manager side depending on the view tag (ref).   So, this should be okay.

thomasmso replied:
If it is not declared for React Native layer to invoke, we should name the method signature properly with argument labels

alhiwatan commented on ios/AppLovinMAXNativeAdView.m:
```diff
@@ -0,0 +1,258 @@
+
+- (void)setAdUnitId:(NSString *)adUnitId
+{
+    if ( self.loadingAd ) return;
```
This is a guard against calling another load before completing the current load.   Hitting a load button repeatedly in a short time caused the demo app stuck.   This is easily reproduced especially in Android.

alhiwatan commented on ios/AppLovinMAXNativeAdView.m:
```diff
@@ -0,0 +1,258 @@
+    self.nativeAdLoader.nativeAdDelegate = self;
+    self.nativeAdLoader.placement = placement;
+    self.nativeAdLoader.customData = customData;
+    for ( NSString* key in extraParameters )
```
Done.

alhiwatan commented on ios/AppLovinMAXNativeAdView.m:
```diff
@@ -0,0 +1,258 @@
+    [self.nativeAdLoader loadAdIntoAdView: [self createNativeAdView]];
+}
+
+- (void)didLoadNativeAd:(MANativeAdView *)nativeAdView forAd:(MAAd *)ad
```
Updated.

alhiwatan commented:
Updated to add AppLovinMaxNativeAdLoader for decoupling the ad loader logic

alhiwatan commented:
@thomasmso , merged and updated.   Please review.

thomasmso commented on ios/AppLovinMAXNativeAdLoader.m:
```diff
@@ -0,0 +1,140 @@
+    
+    self.nativeAdLoaderDelegate = delegate;
+    
+    [self.nativeAdLoader loadAdIntoAdView: [self createNativeAdView]];
```
I still don't understand the hack with `createNativeAdView`. Could you explain in detail?

thomasmso commented on ios/AppLovinMAXNativeAdLoader.m:
```diff
@@ -0,0 +1,140 @@
+    [self.nativeAdLoader loadAdIntoAdView: [self createNativeAdView]];
+}
+
+- (void)didLoadNativeAd:(nullable MANativeAdView *)nativeAdView forAd:(MAAd *)ad
```
Have you tested this on manual and template integrations for native ads?

alhiwatan replied:
Totally forgot about template.   With template, there is no need to do the hack with createNativeAdView.  Removed createNativeAdView.

thomasmso replied:
Ahh ok that makes sense, let's sync up live for more details

alhiwatan commented on example/src/NativeAdViewExample.js:
```diff
@@ -0,0 +1,104 @@
+                              extraParameters={{
+                                'key1': 'value1',
+                                'key2': 'value2',
+                              }}
```
Removed.

alhiwatan commented on ios/AppLovinMAX.m:
```diff
@@ -763,8 +768,15 @@ - (void)didFailToLoadAdForAdUnitIdentifier:(NSString *)adUnitIdentifier withErro
+            [self log: @"invalid adUnitId from %@", [NSThread callStackSymbols]];
+            return;
+        }
     }
```
Updated not to nest.

alhiwatan commented on ios/AppLovinMAXNativeAdViewManager.m:
```diff
@@ -0,0 +1,71 @@
+        UIView *view = viewRegistry[viewTag];
+        if ( !view )
+        {
+            RCTLogError(@"Cannot find UIView with tag %@", viewTag);
```
Updated where the tag is available.

alhiwatan commented on ios/AppLovinMAXNativeAdViewManager.m:
```diff
@@ -0,0 +1,71 @@
+            return;
+        }
+        
+        if ( ![view isKindOfClass:[AppLovinMAXNativeAdView class]] )
```
Updated.

alhiwatan commented on ios/AppLovinMAXNativeAdViewManager.m:
```diff
@@ -0,0 +1,71 @@
+            return;
+        }
+        
+        if ( ![view isKindOfClass:[AppLovinMAXNativeAdView class]] )
```
Updated as well.

alhiwatan commented on ios/AppLovinMAXNativeAdLoader.m:
```diff
@@ -0,0 +1,140 @@
+    else
+    {
+        [AppLovinMAX.shared log: @"callToActionButton is not found in MANativeAdView: %@", self.nativeAdView];
+    }
```
Updated.

alhiwatan commented on ios/AppLovinMAXNativeAdLoader.h:
```diff
@@ -0,0 +1,63 @@
+ */
+- (void)didClickNativeAd:(MAAd *)ad;
+
+@end
```
Removed.

alhiwatan commented on ios/AppLovinMAXNativeAdLoader.h:
```diff
@@ -0,0 +1,63 @@
+/**
+ * Loads a native ad with the specified properties that are set by Javascript.
+ */
+- (void)load:(NSString *)adUnitIdentifier placement:(nullable NSString *)placement customData:(nullable NSString *)customData extraParameters:(nullable NSDictionary *)extraParameters delegate:(id<AppLovinMAXNativeAdLoaderDelegate>)delegate;
```
Separated the arguments with break lines.

alhiwatan commented on ios/AppLovinMAXNativeAdLoader.m:
```diff
@@ -0,0 +1,140 @@
+#import "AppLovinMAX.h"
+#import "AppLovinMAXNativeAdLoader.h"
+
+@interface AppLovinMaxNativeAdLoader ()  <MANativeAdDelegate>
```
Space removed.

alhiwatan commented:
Updated and tested.   Please review.  Now, the ad unit has to be a small/medium template.

alhiwatan commented:
Missed updating the formatting of the consistent return style

thomasmso commented on ios/AppLovinMAXNativeAdViewManager.m:
```diff
@@ -0,0 +1,70 @@
+        
+        if ( ![view isKindOfClass:[AppLovinMAXNativeAdView class]] )
+        {
+            [[AppLovinMAX shared] log: @"Cannot find AppLovinMAXNativeAdViewManager with tag", viewTag];
```
1. Wrong class name in the log,
2. This will crash, since you're adding `viewTag` as an argument, but there is no format specifier in the log.

```suggestion
            [[AppLovinMAX shared] log: @"Cannot find AppLovinMAXNativeAdView with tag %@", viewTag];
```

alhiwatan replied:
Updated.

thomasmso commented on ios/AppLovinMAXNativeAdView.h:
```diff
@@ -0,0 +1,42 @@
+- (instancetype)initWithBridge:(RCTBridge *)bridge;
+
+/**
+ * Loads a native ad, called by Javascript via the View Manager.
```
nit but `S` is usually capitalized when people refer to `JavaScript`. It's free to be lexicographically correct.

alhiwatan replied:
Updated all.

thomasmso reviewed (commented):
Can you set me up with a branch where I can test this locally myself? It's kind of hard to follow the code flow right now

thomasmso commented on android/src/main/java/com/applovin/reactnative/AppLovinMAXNativeAdViewManager.java:
```diff
@@ -0,0 +1,148 @@
+        view.setCustomData( value );
+    }
+
+    @ReactProp(name = "extraParameter")
```
Should be extra parameters right?

alhiwatan replied:
Thanks.  this has to be "extraParameters".   I have verified "extraParameters" on both Android and iOS.

thomasmso reviewed (commented):
Let's also bump iOS SDK version to 11.5.3 (and update changelog to reflect that)
